Add support for wildcard routes: `/foo/bar/{rest:.*}`

ahl opened the pull request:
This will match paths such as `/foo/bar` and `/foo/bar/baz.css`

We can use this to serve static assets such as css, html, and image
files.

Express uses this syntax: `/foo/bar/:rest(.*)`
Dropwizard / Jersey / JAX-RS does: `/foo/bar/{rest:.*}`
Actix does: `/foo/bar/{rest:.*}`

This also adds support for a tag (unpublished = true) in the #[endpoint]
macro to omit certain endpoints from the OpenAPI output. This is both
useful for non-API endpoints and necessary in that OpenAPI doesn't
support any type of multi-segment matching of routes.

david-crespo reviewed (approved):
Looks good to me.

There are a number of spots where we're doing regex-ish logic with loops, which makes wonder if there's a point where it starts to make sense to just use regexes. Based on this tiny benchmark I wrote for `valid_identifier` it doesn't seem like raw performance is a big distinguishing factor, though of course that doesn't say anything about crate bloat or other code organization questions.

https://gist.github.com/david-crespo/1ed0857caa91d37c0a8b9b47f0f124ed

```
loop                    time:   [15.361 ns]                  
loop invalid            time:   [11.025 ns]                          

regex                   time:   [23.126 ns]                   
regex invalid           time:   [15.915 ns]      
```

davepacheco commented on dropshot/src/router.rs:
```diff
@@ -91,21 +91,27 @@ struct HttpRouterNode<Context: ServerContext> {
  * configured.  Each segment may be either a literal string or a variable (the
- * latter indicated by being wrapped in braces.
+ * latter indicated by being wrapped in braces). Variables may consume a single
+ * /-delimited segment or several as defined by a regex (currently only `.*` is
```
There's a couple of things that make me uncomfortable here but these fears might be specious.

One is that we might be conflating "consumes more than one path segment" with "is matched by a regular expression".  I could imagine someone wanting to use regular expressions to match just a segment, as in `/instances/{instance_id:[0-9]+}/status`.  (I'm not really saying we should try to support that, so much as: if we're trying to leave the door open to regular expressions in path segments, it's not obvious to me now that this _shouldn't_ be possible.)

Another is that I think there's some non-trivial work that goes into correctly separating the path segments in HTTP.   This is the kind of stuff I'm thinking about
https://github.com/oxidecomputer/dropshot/blob/e0201cf97431c4326423998dc575b91fc86970ed/dropshot/src/router.rs#L474-L502
Okay, I mostly punted on that, but the goal was to ensure that there's one place where the correct code would go.  These feel like the sort of things that are fine to get wrong for a while and then really hose you when you run into the case where they're wrong.  I'm a little worried that if we allow variables to match multiple segments, then the consumer will try to interpret this path themselves, and it will become really easy to do something that appears to work in some cases but breaks when the input is percent-encoded or something like that.  (If percent-encoding is a real issue, I could imagine this behavior resulting in a vulnerability where somebody thinks they're rejecting requests that contain `/..`, but they're missing requests that contain `'/%2e%2e'`, and this lets people read /etc/passwd or whatever.  Again, this is potentially already broken today because all I have is a TODO, not a correct implementation, but it would suck for this problem to sprawl.)  Does this go against the principle of "make it hard to do the wrong thing"?

We could potentially fix this by saying that instead of a String, you get an array of path segments or something like that.

Am I being too paranoid?

davepacheco commented on dropshot_endpoint/src/lib.rs:
```diff
@@ -150,6 +153,14 @@ fn do_endpoint(
         })
         .unwrap_or_default();
 
+    let visible = if let Some(true) = metadata.unpublished {
```
Is there a test case around visible/unpublished?  (Sorry if I missed it!)

ahl replied:
There is. The openapi tests include an `unpublished` endpoint. We can tell they're successful because no change was required for the output json.

I wasn't sure about the visible vs. unpublished disparity, but I thought that `unpublished` was the right setting for endpoints, but it felt weird to carry the implicit negative into the struct; let me know if you have suggestions for clarity here.

ahl commented:
Based on a discussion with @davepacheco, I changed this around quite a bit. Rather than receiving the path in a `String` this now receives the wild card path in a `Vec<String>`. To do this, I modified the `from_map()` deserializing function to accommodate values that may have an iterator as the value whereas previous only a `String` was allowed.

I observed that consumers could use types for their APIs that were impossible to serialize at runtime when the API endpoint is invoked. To address this, I added more stringent type checking. It's still at runtime, but it happens during `ApiDescription::register()` so very early in the program.

This exposed areas where we were using types that could never work (e.g. `Vec<u16>`) in some tests.

There's still more work to be done here:
- more/better type checking for APIs
- compile-time type checking w/ the `#[endpoint ... ]` macro
- better coordination between path values and path parameter type checking

But this felt like a significant step forward and I wanted to unblock work that will consume the wildcard path matching.

smklein commented on dropshot/src/router.rs:
```diff
@@ -128,12 +136,31 @@ impl PathSegment {
 
-            PathSegment::Varname((&segment[1..segment.len() - 1]).to_string())
+            if let Some(pat) = pat {
+                assert!(
+                    pat == ".*",
+                    "Only the pattern '.*' is currently supported"
+                );
+                PathSegment::VarnameRegEx(var.to_string(), pat.to_string())
```
Reading through more of this PR, it seems like we're doing a lot of special casing for "exactly the wildcard" - not just that this is the only permitted regex, but also things like [checking that no segments exist after the wildcard](https://github.com/oxidecomputer/dropshot/pull/110/files#diff-1d669c288fab8591bd4f3cbcdfb259d761b603299d9a406dd97761344320990aR370).

Seems kinda like those special cases might be hard to undo later - as an example, is it possible to validate that the intersection of N regexes is empty for all possible input paths, to perform a similar collision detection?

ahl replied:
Yes, I'm familiar with the regex crate. I don't follow your second comment.

smklein replied:
ah, let me try to re-phrase.

Basically, when I see structures like this:
https://github.com/oxidecomputer/dropshot/blob/ee7fc26698f1ce680a5dcfd4d0e73280367f6cf5/dropshot/src/router.rs#L94-L101

(specifically the `VariableRest` part)

and checks like this:
https://github.com/oxidecomputer/dropshot/blob/ee7fc26698f1ce680a5dcfd4d0e73280367f6cf5/dropshot/src/router.rs#L297-L305

I interpret that check as: "we'd like to stop someone from registering both `/foo/{bar:.*}` and `/foo/baz`" (as covered by tests below) - but in a more general sense, it seems like we could say "we'd like to prevent callers from registering multiple routes that overlap with one another".

When considering regexes beyond just wildcards, this is where my original question came up - we can maintain checks like this, where we handle only "full wildcard", but making the jump to any other type of regex seems like it'll significantly alter this implementation.

I guess I'm asking what direction the ship is steering towards:

- Do we think pure wildcards will be the only supported regex for the foreseeable future?
- Or do we expect other regexes to be supported? (This led to my question about "finding the intersection between multiple regexes", which seemed like the generalized solution to the "do the routes overlap" question).

ahl replied:
I think I see. I'm going to change `VarnameRegEx` to `VarnameWildcard` to remove the false generalization.

ahl replied:
I posted my second comment before I saw your response. I don't anticipate other regexes being supported any time soon.

ahl commented on dropshot/src/router.rs:
```diff
@@ -153,6 +180,58 @@ impl<'a> From<&'a str> for InputPath<'a> {
+ * Validate the pattern for Rust identifiers:
+ * [a-zA-Z][a-zA-Z0-9_]* | _[a-zA-Z0-9_]+
+ */
+fn valid_identifier(var: &str) -> bool {
+    match var.chars().nth(0) {
```
FWIW the change to supporting unicode went to stable (I think) while this review has been up. You're thinking: `syn::parse_str::<syn::Ident>()`? That's a clever idea. Thanks.

ahl commented on dropshot/src/router.rs:
```diff
@@ -153,6 +180,58 @@ impl<'a> From<&'a str> for InputPath<'a> {
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum VariableValue {
```
I think it should actually be `pub (crate)`, but that pulls a bunch of threads I don't care to pull right now. Yes we can add docs.

ahl commented on dropshot/src/from_map.rs:
```diff
@@ -4,41 +4,62 @@ use paste::paste;
 }
 
 /**
  * Deserializer for BTreeMap<String, String> that interprets the values. It has
```
oof. I sweat I caught another one just like that

ahl commented on dropshot/src/router.rs:
```diff
@@ -757,6 +976,49 @@ mod test {
+        router.insert(new_endpoint(
+            new_handler(),
+            Method::GET,
+            "/some/{more:.*}/{stuff}",
```
Yes, that's what I'd propose. I'll add a test for that; thanks for the suggestion. If that sounds like the wrong approach, please let me know!

ahl replied:
Actually, we currently treat the presence or absence of that trailing slash as identical. It's an area that has some associated `TODO` comments as I think we need to make a determination one way or the other (see `input_path_to_segments`).

smklein replied:
Ack. IMO, still worth the test, even with the final state TBD - seems like behavior that we wouldn't want to alter without knowing about it.

ahl replied:
Fair enough. I'll add a test that succeeds with the final slash and state in the test that it's reasonable to change the behavior although it would be a potentially breaking change.

smklein reviewed (approved):
Looks good to me - to keep up with all other major changes, maybe worth a note in the CHANGELOG?

(I don't believe this is breaking, but it is a visible change)

smklein commented:
FYI, apparently this PR is causing test failures in Omicron, and preventing rolling forward:

https://github.com/oxidecomputer/omicron/pull/154

Trivially repro'd in the Omicron repo by running `./target/debug/nexus omicron-nexus/examples/config.toml  --openapi `, which shows "Aborted (core dumped)".

I'll dig around a bit more, and see if I can root-cause

smklein commented:
... weird, toggling the `panic = "abort"` in `omicron/Cargo.toml` to `panic = "unwind"` avoids the abort altogether.

smklein commented:
Fixed by https://github.com/oxidecomputer/dropshot/pull/122
